vllm/sequence.py

```diff
@@ -486,18 +492,23 @@ def get_output_text_to_return(self, buffer_length: int,
             return self.output_text[last_offset:length]
         return ""
 
-    def get_output_token_ids_to_return(self,
-                                       delta: bool) -> GenericSequence[int]:
+    def get_output_token_ids_to_return(
+            self, delta: bool) -> Union[GenericSequence[int], int]:
         """If delta is True, only new tokens since the last call to
         this method are returned"""
         if not delta:
             return self.get_output_token_ids()
-        length = self.get_output_len()
-        last_offset = self._last_token_ids_offset
-        if last_offset < length:
-            self._last_token_ids_offset = length
-            return self.data._output_token_ids[last_offset:]
-        return ()
+
+        # Optimization for single decode token case
+        # (which is what we have most of the time)
```

Comment:
Did you measure how much difference this makes? What if we use a singleton tuple in this case instead so that the type is still a sequence? I think cpython reuses those and doesn't make a new allocation. Then it shouldn't require special handling.

Reply:
My initial implementation did not had a specific handling for single case and it was a bit slower. I was not large effect, but was about 1-2ms difference in TPOT (of total ~26).